cryptography/hazmat/backends/openssl/backend.py

```diff
@@ -1055,6 +1063,77 @@ def _ec_key_set_public_key_affine_coordinates(self, ctx, x, y):
 
         return ctx
 
+    def generate_dh_parameters(self, generator, key_size):
+        if key_size < 512:
+            raise ValueError("DH key_size must be at least 512 bits")
+
+        dh_param_cdata = self._lib.DH_new()
+        assert dh_param_cdata != self._ffi.NULL
+        dh_param_cdata = self._ffi.gc(dh_param_cdata, self._lib.DH_free)
+
+        res = self._lib.DH_generate_parameters_ex(
+            dh_param_cdata,
+            key_size,
+            generator,
+            self._ffi.NULL
+        )
+        assert res == 1
+
+        return _DHParameters(self, dh_param_cdata)
+
+    def generate_dh_private_key(self, parameters):
+        dh_key_cdata = self._lib.DH_new()
+        assert dh_key_cdata != self._ffi.NULL
+        dh_key_cdata = self._ffi.gc(dh_key_cdata, self._lib.DH_free)
+
+        dh_key_cdata.p = self._lib.BN_dup(parameters._dh_cdata.p)
+        dh_key_cdata.g = self._lib.BN_dup(parameters._dh_cdata.g)
+
+        res = self._lib.DH_generate_key(dh_key_cdata)
+        assert res == 1
+
+        return _DHPrivateKey(self, dh_key_cdata)
+
+    def load_dh_private_numbers(self, numbers):
+        parameter_numbers = numbers.public_numbers.parameter_numbers
+
+        dh_cdata = self._lib.DH_new()
+        assert dh_cdata != self._ffi.NULL
+        dh_cdata = self._ffi.gc(dh_cdata, self._lib.DH_free)
+
+        dh_cdata.p = self._int_to_bn(parameter_numbers.p)
+        dh_cdata.g = self._int_to_bn(parameter_numbers.g)
+        dh_cdata.pub_key = self._int_to_bn(numbers.public_numbers.y)
+        dh_cdata.priv_key = self._int_to_bn(numbers.x)
+
+        codes = self._ffi.new("int[]", 1)
+        res = self._lib.DH_check(dh_cdata, codes)
+        assert res == 1
+        assert codes[0] in (0, 6, 2, 4)
```

Comment:
Why are we using the integers in the assert, instead of binding those
values?

On Tue, Mar 3, 2015 at 6:41 AM, Paul Kehrer <notifications@github.com>
wrote:

> In cryptography/hazmat/backends/openssl/backend.py
> <https://github.com/pyca/cryptography/pull/1279#discussion_r25691575>:
>
> > +    def load_dh_private_numbers(self, numbers):
> > +        parameter_numbers = numbers.public_numbers.parameter_numbers
> > +
> > +        dh_cdata = self._lib.DH_new()
> > +        assert dh_cdata != self._ffi.NULL
> > +        dh_cdata = self._ffi.gc(dh_cdata, self._lib.DH_free)
> > +
> > +        dh_cdata.p = self._int_to_bn(parameter_numbers.p)
> > +        dh_cdata.g = self._int_to_bn(parameter_numbers.g)
> > +        dh_cdata.pub_key = self._int_to_bn(numbers.public_numbers.y)
> > +        dh_cdata.priv_key = self._int_to_bn(numbers.x)
> > +
> > +        codes = self._ffi.new("int[]", 1)
> > +        res = self._lib.DH_check(dh_cdata, codes)
> > +        assert res == 1
> > +        assert codes[0] in (0, 6, 2, 4)
>
> The docs <https://www.openssl.org/docs/crypto/DH_generate_parameters.html>
> say:
>
> "DH_check() validates Diffie-Hellman parameters. It checks that p is a
> safe prime, and that g is a suitable generator. In the case of an error,
> the bit flags DH_CHECK_P_NOT_SAFE_PRIME or DH_NOT_SUITABLE_GENERATOR are
> set in *codes. DH_UNABLE_TO_CHECK_GENERATOR is set if the generator cannot
> be checked, i.e. it does not equal 2 or 5."
>
> The bit flags are in dh.h with these values
>
> /* DH_check error codes */
> # define DH_CHECK_P_NOT_PRIME            0x01
> # define DH_CHECK_P_NOT_SAFE_PRIME       0x02
> # define DH_UNABLE_TO_CHECK_GENERATOR    0x04
> # define DH_NOT_SUITABLE_GENERATOR       0x08
> # define DH_CHECK_Q_NOT_PRIME            0x10
> # define DH_CHECK_INVALID_Q_VALUE        0x20
> # define DH_CHECK_INVALID_J_VALUE        0x40
>
> 0 definitely looks safe, but why would 2, 4, or 6 (2 | 4) be okay?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/pyca/cryptography/pull/1279/files#r25691575>.
>



-- 
"I disapprove of what you say, but I will defend to the death your right to
say it." -- Evelyn Beatrice Hall (summarizing Voltaire)
"The people's good is the highest law." -- Cicero
GPG Key fingerprint: 125F 5C67 DFE9 4084